core/src/main/java/org/mule/runtime/core/internal/processor/strategy/AbstractCachedThreadReactorSinkProvider.java

```diff
@@ -51,9 +72,43 @@ protected void invalidateAll() {
   public FluxSink<CoreEvent> getSink() {
     TransactionCoordination txCoord = TransactionCoordination.getInstance();
     if (txCoord.runningNestedTransaction()) {
-      return sinksNestedTx.get(txCoord.getTransaction(), tx -> createSink());
+      if (sinkIndexEnabled) {
+        return getNestedFluxSinkWrapper(txCoord);
+      } else {
+        return legacySinksNestedTx.get(txCoord.getTransaction(), tx -> new FluxSinkWrapper(createSink()));
+      }
     } else {
-      return sinks.get(currentThread(), t -> createSink());
+      if (sinkIndexEnabled) {
+        return getSimpleFluxSinkWrapper();
+      } else {
+        return legacySinks.get(currentThread(), t -> new FluxSinkWrapper(createSink()));
+      }
+    }
+  }
+
+  private FluxSink<CoreEvent> getNestedFluxSinkWrapper(TransactionCoordination txCoord) {
+    int index = 0;
+    while (true) {
+      FluxSinkWrapper fluxSinkWrapper =
+          sinksNestedTx.get(txCoord.getTransaction().getId() + "-" + index, t -> new FluxSinkWrapper(createSink()));
+      if (fluxSinkWrapper.isBeingUsed()) {
+        index++;
+        continue;
+      }
+      return fluxSinkWrapper;
+    }
+  }
+
+  private FluxSinkWrapper getSimpleFluxSinkWrapper() {
```

Comment:
I see a lot of common code between this method and `getNestedFluxSinkWrapper`.
Couldn't we parameterize it to receive the cache on which to lookup and the key prefix?